### PR TITLE
Setters/substitutions with subpackages

### PR DESCRIPTION
**Pull request by phanimarupaka**
@mortent 

This PR is to introduce a flag, recurse-subpackages in set command, if invoked will recursively run the set command on all the nested subpackages. This PR also changes the current way of setting a setter parameter in the context of subpackages. Currently, we consider subpackages as subfolders and set command sets the values to all the fields with setter comment in the nested subpackages as well. But with this change, the subpackages must contain the setter definitions in them as well for the fields with setter comments inside those packages to be set.

### CODE REVIEW

File: cmd/config/internal/commands/cmdset.go [hunk 1]
```diff
@@ -203,3 +232,23 @@ func (r *SetRunner) perform(c *cobra.Command, args []string) error {
...
 	return nil
 }
+
+// OpenAPIFilePaths returns the openAPI file paths in the root directory
+// returns only file path at top level if RecurseSubPackages is false
+// returns all the file paths in the nested subdirs if RecurseSubPackages is true
+func (r *SetRunner) OpenAPIFilePaths(root string) ([]string, error) {
+	if !r.Set.RecurseSubPackages {
+		if root == "." {
+			return []string{root + "/" + r.Set.OpenAPIFileName}, nil
```
Comment (monopole) [inline cmd/config/internal/commands/cmdset.go:242]: as done below, use `filepath.Join(root, r.Set.OpenAPIFileName)`

Comment (monopole) [reply cmd/config/internal/commands/cmdset.go:242]: i'm confused.

Comment (phanimarupaka) [reply cmd/config/internal/commands/cmdset.go]: filepath.join just returns the file name if the directory root is . in this case, but we want ./fileName as output. Anyways, I refactored this to make it more readable. PTAL again.

File: cmd/config/internal/commands/cmdset.go [hunk 2]
```diff
@@ -126,15 +131,39 @@ func (r *SetRunner) preRunE(c *cobra.Command, args []string) error {
...
+				if len(openAPIPaths) == 1 {
+					return err
+				} else {
+					// print error message and continue if there are multiple packages to set
+					fmt.Fprintf(c.OutOrStdout(), "%s in %s\n", err.Error(), r.Set.ResourcesPath)
+				}
+			} else {
+				fmt.Fprintf(c.OutOrStdout(), "set %d fields in %s\n", count, r.Set.ResourcesPath)
+			}
+			r.Set.Count = 0
```
[the remark this answers is not in the document]
Comment (phanimarupaka) [reply cmd/config/internal/commands/cmdset.go]: Done.

File: cmd/config/internal/commands/cmdset_test.go [hunk 3]
```diff
@@ -1082,3 +1080,71 @@ spec:
...
+		dataset     string
+		packagePath string
+		args        []string
+		expected    string
+	}{
+		{
+			name:    "set-recurse-subpackages",
+			dataset: "dataset1",
+			args:    []string{"namespace", "otherspace", "-R"},
+			expected: `set 1 fields in ${baseDir}/mysql/
```
[the remark this answers is not in the document]
Comment (phanimarupaka) [reply cmd/config/internal/commands/cmdset_test.go]: Done.

File: cmd/config/internal/commands/cmdset.go [hunk 4]
```diff
@@ -126,15 +131,39 @@ func (r *SetRunner) preRunE(c *cobra.Command, args []string) error {
...
-
 	return nil
 }
 
 func (r *SetRunner) runE(c *cobra.Command, args []string) error {
 	if setterVersion == "v2" {
-		count, err := r.Set.Set(r.OpenAPIFile, args[0])
-		fmt.Fprintf(c.OutOrStdout(), "set %d fields\n", count)
-		return handleError(c, err)
+		openAPIFileName, err := ext.OpenAPIFileName()
```
[the remark this answers is not in the document]
Comment (phanimarupaka) [reply cmd/config/internal/commands/cmdset.go:138]: ext is is cmd/config pkg and we want to pass it to kyaml. So doing it this way.

File: cmd/config/internal/commands/cmdset.go [hunk 5]
```diff
@@ -126,15 +130,53 @@ func (r *SetRunner) preRunE(c *cobra.Command, args []string) error {
...
+				Name:               r.Set.Name,
+				Value:              r.Set.Value,
+				ListValues:         r.Set.ListValues,
+				Description:        r.Set.Description,
+				SetBy:              r.Set.SetBy,
+				Count:              0,
+				OpenAPIPath:        filepath.Join(resourcesPath, openAPIFileName),
+				OpenAPIFileName:    openAPIFileName,
+				ResourcesPath:      resourcesPath,
+				RecurseSubPackages: r.Set.RecurseSubPackages,
```
Comment (mortent) [inline cmd/config/internal/commands/cmdset.go:164]: Since we are creating a new FieldSetter for each package/subpackage, why would we need to pass this along? It seems like a FieldSetter only needs to handle one package.

File: cmd/config/internal/commands/test/testdata/dataset2/mysql/Krmfile [hunk 6]
```diff
@@ -0,0 +1,6 @@
+apiVersion: krm.dev/v1alpha1
```
Comment (mortent) [inline cmd/config/internal/commands/test/testdata/dataset2/mysql/Krmfile:1]: Can we make the testdata here a bit more understandable to others? At least name the different datasets with names that actually describe what they are meant for.

File: kyaml/setters2/settersutil/settercreator.go [hunk 7]
```diff
@@ -165,3 +173,36 @@ func validateSchemaTypes(sc *spec.Schema) error {
 	}
 	return nil
 }
+
+func (c SetterCreator) validateSetterInfo() error {
+	openapi.ResetOpenAPI()
```
Comment (mortent) [inline kyaml/setters2/settersutil/settercreator.go:178]: We can't reset the openapi schema here. If the user has read the schema from a file or fetched it from the cluster, this will reset it with the builtin schema

File: kyaml/pathutil/pathutil.go [hunk 8]
```diff
@@ -9,20 +9,27 @@ import (
...
+		func(filePath string, info os.FileInfo, err error) error {
 			if err != nil {
 				return err
 			}
-			if strings.HasSuffix(path, fileName) {
-				if root == "." {
-					path = root + "/" + path
-				}
-				res = append(res, path)
+			if strings.HasSuffix(filePath, fileName) {
```
Comment (mortent) [inline kyaml/pathutil/pathutil.go:31]: I think we should use a function from the `filepath` package here to get the name of the file and then check if the filename match what we are looking for.

File: kyaml/pathutil/pathutil.go [hunk 9]
```diff
@@ -9,20 +9,27 @@ import (
...
+		// else return empty list
+		_, err := os.Stat(filepath.Join(root, fileName))
+		if !os.IsNotExist(err) {
+			res = append(res, root)
+		}
+		return res, nil
+	}
 	err := filepath.Walk(root,
-		func(path string, info os.FileInfo, err error) error {
+		func(filePath string, info os.FileInfo, err error) error {
```
Comment (mortent) [inline kyaml/pathutil/pathutil.go:27]: Use a different name for the first parameter here. Having a variable named `filePath` and an imported package called `filepath` is confusing.

File: cmd/config/internal/commands/cmdcreatesubstitution_test.go [hunk 10]
```diff
@@ -419,3 +421,78 @@ spec:
...
+			packagePath: "mysql/storage",
+			args:        []string{"image-tag", "--field-value", "storage:1.7.9", "--pattern", "${image}:${tag}"},
+			expected:    `created substitution image-tag in package ${baseDir}/mysql/storage`,
+		},
+		{
+			name:        "create-subst-already-exists",
+			dataset:     "dataset1",
+			packagePath: "mysql",
+			args:        []string{"image-tag", "--field-value", "mysql:1.7.9", "--pattern", "${image}:${tag}", "-R"},
+			expected: `substitution with name image-tag already exists in package ${baseDir}/mysql
```
[the remark this answers is not in the document]
Comment (phanimarupaka) [reply cmd/config/internal/commands/cmdcreatesubstitution_test.go]: Done.

File: cmd/config/internal/commands/cmdset.go (hunk 5 above)
Comment (phanimarupaka) [reply cmd/config/internal/commands/cmdset.go:164]: So that the behavior is consistent both for cli and api. It is not being used now at the api but wanted to be consistent.

File: cmd/config/internal/commands/test/testdata/dataset2/mysql/Krmfile (hunk 6 above)
Comment (phanimarupaka) [reply cmd/config/internal/commands/test/testdata/dataset2/mysql/Krmfile]: Done.

File: kyaml/pathutil/pathutil.go (hunk 8 above)
Comment (phanimarupaka) [reply kyaml/pathutil/pathutil.go]: Done.

File: kyaml/pathutil/pathutil.go (hunk 9 above)
Comment (phanimarupaka) [reply kyaml/pathutil/pathutil.go]: Done.

File: kyaml/setters2/settersutil/settercreator.go (hunk 7 above)
Comment (phanimarupaka) [reply kyaml/setters2/settersutil/settercreator.go:178]: I expected this comment. Setter commands in their current form doesn't use k8s schema. They only need setters schema. So had to do this so that only the setter schema for the current package is only loaded for validation.

File: cmd/config/internal/commands/cmdcreatesetter.go [hunk 11]
```diff
@@ -212,7 +179,49 @@ func (r *CreateSetterRunner) processSchema() error {
...
+		resourcePackagesPaths, err := pathutil.DirsWithFile(args[0], openAPIFileName, r.CreateSetter.RecurseSubPackages)
+		if err != nil {
+			return err
+		}
+
+		if len(resourcePackagesPaths) == 0 {
+			return errors.Errorf("unable to find %s in package %s", r.CreateSetter.OpenAPIFileName, args[0])
+		}
+		for _, resourcesPath := range resourcePackagesPaths {
+			r.CreateSetter = settersutil.SetterCreator{
```
[the remark this answers is not in the document]
Comment (phanimarupaka) [reply cmd/config/internal/commands/cmdcreatesetter.go:197]: We currently combine the cli and api objects. We should clean this up by separating them. We can do that as part of future PRs to clean this up all across.

File: kyaml/setters2/settersutil/settercreator.go (hunk 7 above)
Comment (phanimarupaka) [reply kyaml/setters2/settersutil/settercreator.go]: Fixed it. Added a new method to delete schema present in file from global schema so that for every package only the setter schema related to it is loaded and previous package setters schema is cleaned up.

File: cmd/config/internal/commands/cmdset.go (hunk 5 above)
Comment (mortent) [reply cmd/config/internal/commands/cmdset.go:164]: Yeah, I'm ok with leaving this for now. But we should clean this up in a follow-up PR.

File: kyaml/kio/ignorefilesmatcher.go [hunk 12]
```diff
@@ -30,14 +30,14 @@ import (
 //   package contains a pattern that ignores the directory foo, if foo is a
 //   subpackage, it will still be included if the IncludeSubpackages property
 //   is set to true
-type ignoreFilesMatcher struct {
+type IgnoreFilesMatcher struct {
```
Comment (mortent) [inline kyaml/kio/ignorefilesmatcher.go:33]: Why is this needed? The comment should be updated with the new name of the type.

File: cmd/config/internal/commands/cmdcreatesubstitution.go [hunk 13]
```diff
@@ -49,49 +51,58 @@ type CreateSubstitutionRunner struct {
...
+			fmt.Fprintf(c.OutOrStdout(), "created substitution %q in package %q\n", r.CreateSubstitution.Name, r.CreateSubstitution.ResourcesPath)
+		}
 
-	setter, _ := openapi.Resolve(&ref)
-	// if setter already exists with input substitution name, throw error
-	if setter != nil {
-		return errors.Errorf(fmt.Sprintf("setter with name %s already exists, "+
-			"substitution and setter can't have same name", r.CreateSubstitution.Name))
+		// Delete schema present in openAPI file for current package
+		if err := openapi.DeleteSchemaInFile(r.CreateSubstitution.OpenAPIPath); err != nil {
```
Comment (mortent) [inline cmd/config/internal/commands/cmdcreatesubstitution.go:99]: It seems like this should probably be done with a `defer` call?
